content-services/content-sources-backend
# Fixes 4548: add public upload create and upload chunk apis

## Summary

- Adds the public APIs for creating an upload and uploading a chunk: `POST /repositories/uploads/` and `POST /repositories/uploads/:upload_uuid/upload_chunk/`
- Modifies the existing addUploads API to use either an upload UUID or an upload href
- Modifies the python script for testing the upload process to use either the public or internal APIs

## Testing steps

- Create an upload repository with snapshot set to true
- Use the ./scripts/uploads.py script to use the public APIs to create an upload, upload the chunks, and add the uploads:
`python3 uploads.py <repo_uuid> <rpm1 rpm2 ...> --api public` (the same script should still be able to be used with the internal APIs: `python3 uploads.py <repo_uuid> <rpm1 rpm2 ...> --api internal`)
- Fetch the repository and verify the package count (should be 1)
- Fetch the repository's rpms and the snapshot's rpms and verify they include the one you just uploaded
- Access the pulp content via the pulp URL and verify the expected rpm is there
- Upload the same rpm and verify no new snapshot is created
- Upload a different rpm and verify a new snapshot is created with that rpm added
- For QE, the upload chunk API in particular may be difficult to test and I may need to change some things so this works with generated client. 

Here are some example commands in case it's helpful:

```
Create upload request:
curl -X POST 'localhost:8000/api/content-sources/v1.0/repositories/uploads/' \
-H 'x-rh-identity: <identity_header>' \
-H 'Content-Type: application/json' \
-d '{    
    "size": 7036
}'

IQE shell:
app.content_sources.rest_client.repositories_api.create_upload({"size": 7036})

Response:
{
    "upload_uuid": "01916bde-8e8a-76a6-aa34-17acdfb86ceb",
    "created": "2024-08-19T18:20:33.290458Z",
    "last_updated": "2024-08-19T18:20:33.290468Z",
    "size": 7036
}

Upload chunk request:
curl -X POST 'localhost:8000/api/content-sources/v1.0/repositories/uploads/01916bde-8e8a-76a6-aa34-17acdfb86ceb/upload_chunk/' \
-H 'x-rh-identity:  <identity_header>' \
-H 'Content-Range: bytes 0-7035/*' \
-F 'file=@"/Users/bhouse/rpms/rpm_chunkaa"' \
-F 'sha256="7e380bd4c000378f407c1a3663dedfda01248564fd7d4e33d56a1c691046bce1"'

IQE shell:
with open('/Users/bhouse/rpms/rpm_chunkaa', 'rb') as file:
   response = app.content_sources.rest_client.repositories_api.upload_chunk('01916bde-8e8a-76a6-aa34-17acdfb86ceb', 'bytes 0-7035/*', file, '7e380bd4c000378f407c1a3663dedfda01248564fd7d4e33d56a1c691046bce1')
response

Response:
{
    "upload_uuid": "01916bde-8e8a-76a6-aa34-17acdfb86ceb",
    "created": "2024-08-19T18:20:33.290458Z",
    "last_updated": "2024-08-19T18:20:33.290468Z",
    "size": 7036
}

Add uploads request:
curl -X POST 'localhost:8000/api/content-sources/v1.0/repositories/8a578521-50be-48df-b029-d7f240aa7dfa/add_uploads/' \
-H 'x-rh-identity: <identity_header>' \
-H 'Content-Type: application/json' \
-d '{
    "uploads": [
        {
            "uuid": "01916bde-8e8a-76a6-aa34-17acdfb86ceb",
            "sha256": "7e380bd4c000378f407c1a3663dedfda01248564fd7d4e33d56a1c691046bce1"
        }
    ]
}'

IQE shell:
body = { 'uploads': [{'uuid':'01916bde-8e8a-76a6-aa34-17acdfb86ceb', 'sha256':'7e380bd4c000378f407c1a3663dedfda01248564fd7d4e33d56a1c691046bce1'}], 'artifacts': [] }
app.content_sources.rest_client.repositories_api.add_upload('8a578521-50be-48df-b029-d7f240aa7dfa', body)


Response:
{
    "uuid": "29125d02-d11c-4d44-973d-878c9b963d64",
    "status": "pending",
    "created_at": "2024-08-22T11:55:13-04:00",
    "ended_at": "",
    "error": "",
    "org_id": "17684632",
    "type": "add-uploads-repository",
    "repository_name": "test",
    "repository_uuid": "8a578521-50be-48df-b029-d7f240aa7dfa",
    "dependencies": [],
    "dependents": []
}

```
